syntax = "proto3";

import "google/protobuf/empty.proto";

service Process {
  rpc Initialize(InitializeRequest) returns (google.protobuf.Empty) {}
  rpc Clear(google.protobuf.Empty) returns (google.protobuf.Empty) {}
  rpc SetPredecessorIP(SetPredecessorIPRequest) returns (google.protobuf.Empty) {}
  rpc SetRole(ProcessRole) returns (google.protobuf.Empty) {}
  rpc GetNumericalDeviation(NumericalDeviationRequest) returns (NumericalDeviation) {}
  rpc Reconcile(ReconcileRequest) returns (google.protobuf.Empty) {}
  rpc RawWrite(RawWriteRequest) returns (google.protobuf.Empty) {}
  rpc Write(WriteRequest) returns (google.protobuf.Empty) {}
  rpc Read(ReadRequest) returns (ReadResponse) {}
  rpc DataStatus(google.protobuf.Empty) returns (StatusResponse) {}
  rpc ListBooks(google.protobuf.Empty) returns (BookList) {}
}

message InitializeRequest {
  string processID = 1;
  string predecessorIP = 3;
  string successorIP = 4;
  string tailIP = 5;
  string headIP = 6;
  int32 role = 7;
}

message SetPredecessorIPRequest {
  string processID = 1;
  string ip = 2;
}

message ProcessRole {
  string processID = 1;
  int32 role = 2;
}

message NumericalDeviationRequest {
  string processID = 1;
}

message NumericalDeviation {
  string processID = 1;
  int32 deviation = 2;
}

message ReconcileRequest {
  string sourceProcessID = 1;
  string targetProcessID = 2;
  string targetIP = 3;
}

message RawWriteRequest {
  string processID = 1;
  string key = 2;
  string value = 3;
}

message WriteRequest {
  string key = 1;
  float value = 2;
  uint32 timeout = 3;
}

message ReadRequest {
  string key = 1;
}

message ReadResponse {
  float value = 1;
  bool success = 2;
}

message StatusResponse {
  map<string, string> status = 1;
}

message BookList {
  map<string, float> books = 1;
}